File: app/src/main/java/com/nao4j/subtrol/security/UserAuthentication.java
package com.nao4j.subtrol.security;

import com.nao4j.subtrol.dto.UserCredentials;
import lombok.Data;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;

import java.util.Collection;

import static java.util.stream.Collectors.toList;

@Data
public class UserAuthentication implements Authentication {

    private final UserCredentials user;
    private boolean authenticated = true;

    @Override
    public String getName() {
        return user.getId();
    }

    @Override
    public Collection<GrantedAuthority> getAuthorities() {
        return user.getRoles().stream().map(SimpleGrantedAuthority::new).collect(toList());
    }

    @Override
    public String getCredentials() {
        return user.getPassword();
    }

    @Override
    public UserCredentials getDetails() {
        return user;
    }

    @Override
    public String getPrincipal() {
        return user.getId();
    }

    @Override
    public boolean isAuthenticated() {
        return authenticated;
    }

    @Override
    public void setAuthenticated(final boolean authenticated) {
        this.authenticated = authenticated;
    }

}

File: src/main/kotlin/com/nao4j/subtrol/security/UserAuthentication.kt
package com.nao4j.subtrol.security

import com.nao4j.subtrol.dto.UserCredentials
import org.springframework.security.core.Authentication
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority

class UserAuthentication(private val user: UserCredentials) : Authentication {

    private var authenticated = true

    override fun getName(): String {
        return user.email
    }

    override fun getAuthorities(): Collection<GrantedAuthority> {
        return user.roles.map { SimpleGrantedAuthority(it) }
    }

    override fun getCredentials(): Any {
        return user.password
    }

    override fun getDetails(): UserCredentials {
        return user
    }

    override fun getPrincipal(): Any {
        return user.email
    }

    override fun isAuthenticated(): Boolean {
        return authenticated
    }

    override fun setAuthenticated(authenticated: Boolean) {
        this.authenticated = authenticated
    }

}
